packages/web-components/fast-foundation/src/virtual-list/virtual-list.ts

```diff
@@ -0,0 +1,877 @@
+        }
+
+        return {
+            start: itemIndex * this.itemSize,
+            end: itemIndex * this.itemSize + this.itemSize,
+            size: this.itemSize,
+        };
+    };
+
+    /**
+     * starts observing the items array
+     */
+    private observeItems(): void {
+        if (!this.items) {
+            return;
+        }
+
+        this.unobserveItems();
+
+        // TODO:  we don't use splices calculated by array change events
+        // look for cheaper observer implementation later
+
+        const newObserver = (this.itemsObserver = Observable.getNotifier(this.items));
+        newObserver.subscribe(this);
+    }
+
+    /**
+     * stops observing the items array
+     */
+    private unobserveItems(): void {
+        this.itemsObserver?.unsubscribe(this);
+        this.itemsObserver = null;
+    }
+
+    /**
+     * The items list has mutated
+     *
+     * @internal
+     */
+    public handleChange(source: any, splices: Splice[]): void {
+        const itemsLength = this.items.length;
+        const firstRenderedIndex = Math.min(this.firstRenderedIndex, itemsLength - 1);
+        const lastRenderedIndex = Math.min(this.lastRenderedIndex, itemsLength - 1);
+
+        const newVisibleItems = this.items.slice(
+            firstRenderedIndex,
+            lastRenderedIndex + 1
+        );
+
+        this.visibleItems.splice(0, this.visibleItems.length, ...newVisibleItems);
+
+        this.updateDimensions();
+        this.requestPositionUpdates();
+    }
+
+    /**
+     * get position updates
+     */
+    protected requestPositionUpdates(): void {
+        if (!this.virtualizationEnabled) {
+            this.updateVisibleItems();
+            return;
+        }
+        if (this.pendingPositioningUpdate) {
+            this.finalUpdateNeeded = true;
+            return;
+        }
+        this.finalUpdateNeeded = false;
+        this.pendingPositioningUpdate = true;
+
+        VirtualList.intersectionService.requestPosition(
+            this.containerElement,
+            this.handleIntersection
+        );
+        VirtualList.intersectionService.requestPosition(
+            this.viewportElement,
+            this.handleIntersection
+        );
+    }
+
+    /**
+     * request reset
+     */
+    protected reset(): void {
+        if (this.pendingReset) {
+            return;
+        }
+
+        this.pendingReset = true;
+
+        DOM.queueUpdate(() => {
+            this.doReset();
+        });
+    }
+
+    /**
+     * execute reset
+     */
+    private doReset(): void {
+        this.pendingReset = false;
+        this.cancelPendingPositionUpdates();
+        this.observeItems();
+        this.updateDimensions();
+    }
+
+    /**
+     * initialize repeat behavior for visible items
+     */
+    private initializeRepeatBehavior(): void {
+        if (this.itemsRepeatBehavior !== null) {
+            return;
+        }
+
+        this.itemsRepeatBehavior = new RepeatDirective(
```

Comment:
No need to change anything right now but flagging for awareness: I think this pattern could cause issues with SSR because the directive isn't actually part of the template and is executed imperatively; I think this *could* throw or possible just omit any DOM emitted by this repeat directive. If there is a way to keep this directive in the template itself that would mitigate the issue. I'll keep this scenario in mind and see if there is a way to better support this scenario in the SSR infrastructure.

Reply:
We already do this in datagrid - https://github.com/microsoft/fast/blob/793975df25def4178ab56a26ab0fd41c7f2fcf7f/packages/web-components/fast-foundation/src/data-grid/data-grid.ts#L316

Reply:
Note that the reason for this approach is to have the repeat render elements in the light dom.  Maybe addressing that is the fix for this case?